### PR TITLE
chore: add example how to use data loaders with suspendable functions

### PR DESCRIPTION
**Pull request by samuelAndalon**
### :pencil: Description
adding an example of how to use DataLoaders and keep using suspendable functions

### :link: Related Issues
https://github.com/ExpediaGroup/graphql-kotlin/issues/986

### CODE REVIEW

**Comment by samuelAndalon** [comment]
if everything looks ok, i will also add this to the 6.x.x docs

File: website/docs/server/data-loader/data-loader.md [hunk 1]
```diff
@@ -97,8 +97,97 @@ class User(val id: ID) {
...
-:::info
-Given that `graphql-java` relies on `CompletableFuture`s for scheduling and asynchronous execution of `DataLoader` calls,
-currently we don't provide any native support for `DataLoader` pattern using coroutines. Instead, return
+## DataLoaders and Coroutines
+
+`graphql-java` relies on `CompletableFuture` for scheduling and execute asynchronous field resolvers (aka data fetchers),
+`graphql-java` deliberately aims for near zero dependencies which means no Kotlin / WebFlux / RxJava.
+
+Similar to `asynchronous` field resolvers (aka data fetcher), `DataLoader` pattern implementation in `graphql-java` works with `CompletableFuture`,
+and because of the listed `graphql-java` constrains we don't provide any native support for `DataLoader` pattern using suspendable functions. Instead, return
```
Comment (dariuszkuc) [inline website/docs/server/data-loader/data-loader.md:106]: IMHO previous version was clearer - I don't think adding the section about the data fetchers is needed and it brings confusion, i.e. yes `graphql-java` does use CF for data fetchers but we provide interop with coroutines but also `graphql-java` uses CF for data loaders but we don't provide support for it?

Adding details on why it works on data fetchers but not on data loaders might be too much so I'd just keep the previous comment.

File: website/docs/server/data-loader/data-loader.md [hunk 2]
```diff
@@ -97,8 +97,97 @@ class User(val id: ID) {
...
+            coroutineScope.future {
+                userService.getUsers(ids)
+            }
+        }
+}
+```
+
+There are some things going on here:
+
+1. We define the `UserDataLoader` with name "UserDataLoader"
```
Comment (dariuszkuc) [inline website/docs/server/data-loader/data-loader.md:166]: Maybe add comments to the code above to indicate what are you referring to?

i.e. something along those lines
```kotlin
class UserDataLoader : KotlinDataLoader<ID, User> { // 1
```

File: website/docs/server/data-loader/data-loader.md [hunk 3]
```diff
@@ -97,8 +97,97 @@ class User(val id: ID) {
...
+2. The `getLoader()` method returns a `DataLoader<Int, User>`, which `BatchLoader` function should return a `List<User>`
```
Comment (dariuszkuc) [inline website/docs/server/data-loader/data-loader.md:167]: there is no `getLoader()` method, did you mean `getDataLoader()`?

File: website/docs/server/data-loader/data-loader.md [hunk 4]
```diff
@@ -97,8 +97,97 @@ class User(val id: ID) {
...
+3. Given that we **don't want** to change our `UserService` async model that is using coroutines, we need a `CoroutineScope`, [which is conveniently available](../../schema-generator/execution/async-models/#coroutines) in the `GraphQLContext`, which is provided in the `DataFetchingEnvironment` and accessible with the [getGraphQLContext()](https://github.com/ExpediaGroup/graphql-kotlin/blob/master/executions/graphql-kotlin-dataloader-instrumentation/src/main/kotlin/com/expediagroup/graphql/dataloader/instrumentation/extensions/BatchLoaderEnvironmentExtensions.kt#L43) extension function.
```
Comment (dariuszkuc) [inline website/docs/server/data-loader/data-loader.md:168]:
```suggestion
3. Given that we **don't want** to change our `UserService` async model that is using coroutines, we need a `CoroutineScope`, [which is conveniently available](../../schema-generator/execution/async-models/#coroutines) in the `GraphQLContext` and accessible through [`DataFetchingEnvironment#getGraphQLContext()`](https://github.com/ExpediaGroup/graphql-kotlin/blob/master/executions/graphql-kotlin-dataloader-instrumentation/src/main/kotlin/com/expediagroup/graphql/dataloader/instrumentation/extensions/BatchLoaderEnvironmentExtensions.kt#L43) extension function.
```

File: website/docs/server/data-loader/data-loader.md [hunk 5]
```diff
@@ -97,8 +97,97 @@ class User(val id: ID) {
...
+4. After retrieving the `CoroutineScope` from the `batchLoaderEnvironment` we will be able to execute the `userService.getUsers(ids)` suspendable function.
+5. We interoperate the suspendable function result to a `CompletableFuture` using [coroutineScope.future](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-jdk8/kotlinx.coroutines.future/future.html)
+
+Finally, the only thing that we need to change is the `user` field resolver, to not be suspendable and
+just return the `CompletableFuture<User>` that the `DataLoader`, make sure to pass the `dataFetchingEnvironment` as `keyContext` which is the second argument of `DataLoader.load`
```
Comment (dariuszkuc) [inline website/docs/server/data-loader/data-loader.md:173]:
```suggestion
Finally, we need to update `user` field resolver, to return the `CompletableFuture<User>` from the invoked `DataLoader`. Make sure to update method signature to also accept the `dataFetchingEnvironment` as you need to pass it to `DataLoader#load` method to be able to execute the request in appropriate coroutine scope.
```

File: website/docs/server/data-loader/data-loader.md [hunk 6]
```diff
@@ -97,8 +97,101 @@ class User(val id: ID) {
...
 ```
 
-:::info
-Given that `graphql-java` relies on `CompletableFuture`s for scheduling and asynchronous execution of `DataLoader` calls,
-currently we don't provide any native support for `DataLoader` pattern using coroutines. Instead, return
-the `CompletableFuture` directly from your `DataLoader`s. See issue [#986](https://github.com/ExpediaGroup/graphql-kotlin/issues/986).
-:::
+## DataLoaders and Coroutines
+
+`graphql-java` relies on `CompletableFuture`s for scheduling and asynchronously execute GraphQL operations.
```
Comment (dariuszkuc) [inline website/docs/server/data-loader/data-loader.md:102]:
```suggestion
`graphql-java` relies on `CompletableFuture`s for scheduling and asynchronously executing GraphQL operations.
```

File: website/docs/server/data-loader/data-loader.md [hunk 7]
```diff
@@ -97,8 +97,101 @@ class User(val id: ID) {
...
+While we can provide native support for coroutines for data fetchers (aka field resolvers) because they are resolved
+independently, we cannot easily provide native support for the `DataLoader` pattern as it relies
+on `CompletableFuture` and we cannot update it to use coroutines state machine logic without fully rewriting
+GraphQL Java execution engine.
```
Comment (dariuszkuc) [inline website/docs/server/data-loader/data-loader.md:106]:
```suggestion
on `CompletableFuture` state machine internals and we cannot update it to use coroutines without fully rewriting
GraphQL Java execution engine.
```

File: website/docs/server/data-loader/data-loader.md [hunk 8]
```diff
@@ -97,8 +97,101 @@ class User(val id: ID) {
...
+
+If you would like to use `DataLoader` pattern in your project, you have to update your data fetchers (aka field resolvers) to return
+`CompletableFuture` from the invoked `DataLoader` instead of marking it a `suspend`.
```
Comment (dariuszkuc) [inline website/docs/server/data-loader/data-loader.md:109]:
```suggestion
If you would like to use `DataLoader` pattern in your project, you have to update your data fetchers (aka field resolvers) to return
`CompletableFuture` from the invoked `DataLoader`.
```

File: website/versioned_docs/version-6.x.x/server/data-loader/data-loader.md [hunk 9]
```diff
@@ -97,8 +97,101 @@ class User(val id: ID) {
...
+
+When we execute the above query, we will end up calling `UserService#getUser` twice which will result in two independent
+downstream service/database calls. This problem is called N+1 problem. By using `DataLoader` pattern,
+we can solve this problem and only make a single downstream request/query.
+
+Lets create the `UserDataLoader`:
+
+```kotlin
+class UserDataLoader : KotlinDataLoader<ID, User> {  // 1
+    override val dataLoaderName = "UserDataLoader"
```
Comment (tapaderster) [inline website/versioned_docs/version-6.x.x/server/data-loader/data-loader.md:153]: should this line be `// 1` instead?

**Review by tapaderster — APPROVED** [review]

**Pull request merged**